feat: what's new for 1.7.0 (#3000)
JetBrains/kotlin-web-site: docs/topics/whatsnew17.md

projects, and none of compiler plugins, including [kapt](kapt.md), work with it.

Our benchmarks show some outstanding results on our internal projects:

| Project       | Current Kotlin compiler performance | New K2 Kotlin compiler performance | Performance boost |
|---------------|-------------------------------------|------------------------------------|-------------------|
| Kotlin        | 2.2 KLOC/s                          | 4.8 KLOC/s                         | ~ x2.2            |
| YouTrack      | 1.8 KLOC/s                          | 4.2 KLOC/s                         | ~ x2.3            |
| IntelliJ IDEA | 1.8 KLOC/s                          | 3.9 KLOC/s                         | ~ x2.2            |
| Space         | 1.2 KLOC/s                          | 2.8 KLOC/s                         | ~ x2.3            |

> The KLOC/s performance numbers stand for the number of thousands of lines of code that the compiler processes per
> second.
>
> {type="tip"}

You can check out the performance boost on your JVM projects and compare it with the results of the old compiler. To
enable the Kotlin K2 compiler, use the following compiler option:

```bash
-Xuse-k2
```

> Also, the K2 compiler includes a number of bugfixes. You can use this list as a
> reference: [https://youtrack.jetbrains.com/issues/KT?q=tag:%20fixed-in-frontend-ir%20sort%20by:%20Priority,%20votes,%20updated](https://youtrack.jetbrains.com/issues/KT?q=tag:%20fixed-in-frontend-ir%20sort%20by:%20Priority,%20votes,%20updated).
> Please note that even issues with State: Open from this list are in fact fixed in K2.

Reviewer: ```suggestion
Also, the K2 compiler [includes a number of bugfixes](https://youtrack.jetbrains.com/issues/KT?q=tag:%20fixed-in-frontend-ir%20sort%20by:%20Priority,%20votes,%20updated).
Please note that even issues with **State: Open** from this list are in fact fixed in K2.
```

@JvmInline
value class BarWrapper(val bar: Bar): Bar by bar

fun main() {
    val bw = BarWrapper(object: Bar {})
    println(bw.foo())
}
```

### Underscore operator for type arguments

Kotlin 1.7.0 introduces an underscore operator, `_`, for type arguments. You can use it to automatically infer a type
argument when other types are specified:

```kotlin
abstract class SomeClass<T> {
    abstract fun execute(): T
}

class SomeImplementation : SomeClass<String>() {
    override fun execute(): String = "Test"
}

class OtherImplementation : SomeClass<Int>() {
    override fun execute(): Int = 42
}

object Runner {
    inline fun <reified S: SomeClass<T>, T> run(): T {
        return S::class.java.getDeclaredConstructor().newInstance().execute()
    }
}

fun main() {
    // T is inferred as String because SomeImplementation derives from SomeClass<String>
    val s = Runner.run<SomeImplementation, _>()
    assert(s == "Test")

    // T is inferred as Int because OtherImplementation derives from SomeClass<Int>
    val n = Runner.run<OtherImplementation, _>()
    assert(n == 42)
}
```

> You can use the underscore operator in any position in the variables list to infer a type argument.
>

> > {type="note"}

Reviewer: ```suggestion
{type="note"}
```

### Compiler performance optimizations

Kotlin 1.7.0 introduces performance improvements for the Kotlin/JVM compiler. According to our benchmarks, compilation
time has been [reduced by 10% on average](https://youtrack.jetbrains.com/issue/KT-48233/Switching-to-JVM-IR-backend-increases-compilation-time-by-more-t#focus=Comments-27-6114542.0-0)
compared to Kotlin 1.6.0. Projects with lots of usages of inline functions, for
example, [projects using `kotlinx.html`](https://youtrack.jetbrains.com/issue/KT-51416/Compilation-of-kotlinx-html-DSL-should-still-be-faster),
will compile faster thanks to the improvements to the bytecode postprocessing.

### New compiler option: -Xjdk-release

Kotlin 1.7.0 presents a new compiler option, `-Xjdk-release`. This option is similar to
the [javac's command-line `--release` option](http://openjdk.java.net/jeps/247). The `-Xjdk-release` option controls the
target bytecode version and limits the API of the JDK in the classpath to the specified Java version. For
example, `kotlinc -Xjdk-release=1.8` won't allow referencing `java.lang.Module` even if the JDK in the dependencies is
version 9 or higher.

> This option works only for OpenJDK distributions.
>

> > {type="note"}

Reviewer: ```suggestion
{type="note"}
```

* [A new approach to incremental compilation](#a-new-approach-to-incremental-compilation)
* [New build reports for tracking compiler performance](#build-reports-for-kotlin-compiler-tasks)
* [Changes to the minimum supported versions of Gradle and the Android Gradle plugin](#bumping-minimum-supported-versions)
* [Support for Gradle plugin variants](#support-for-gradle-plugin-variants)
* [Updates in the Kotlin Gradle plugin API](#updates-in-the-kotlin-gradle-plugin-api)
* [Availability of the sam-with-receiver plugin via the plugins API](#the-sam-with-receiver-plugin-is-available-via-the-plugins-api)
* [Changes in compile tasks](#changes-in-compile-tasks)
* [New statistics of generated files by each annotation processor in kapt](#statistics-of-generated-files-by-each-annotation-processor-in-kapt)
* [Deprecation of the kotlin.compiler.execution.strategy system property](#deprecation-of-the-kotlin-compiler-execution-strategy-system-property)
* [Removal of deprecated options, methods, and plugins](#removal-of-deprecated-options-methods-and-plugins)

### A new approach to incremental compilation

> > The new approach to incremental compilation is [Experimental](components-stability.md). It may be dropped or
> > changed at any time.
> 
> > Opt-in is required (see the details below). We encourage you to Uuse it only for evaluation purposes, and. Wwe would
> > appreciate your feedback in [YouTrack](https://youtrack.jetbrains.com/issues/KT).
> 
> >
> {type="warning"}

Reviewer: ```suggestion
> The new approach to incremental compilation is [Experimental](components-stability.md). It may be dropped or changed at any time.
> Opt-in is required (see the details below). We encourage you to use it only for evaluation purposes, and we would
> appreciate your feedback in [YouTrack](https://youtrack.jetbrains.com/issues/KT).
>
{type="warning"}
```

Gradle. Each variant will have implementations for Gradle features from a corresponding release. The latest variant will
support the widest Gradle feature set. With this approach, we can extend support for older Gradle versions with limited
functionality.

Currently, there are only two variants of the Kotlin Gradle plugin:

* `main` for Gradle versions 6.7.1–6.9.2
* `gradle70` for Gradle versions 7.0 and higher

In future Kotlin releases, we may add more.

To check which variant your build uses, enable
the [`--info` log level](https://docs.gradle.org/current/userguide/logging.html#sec:choosing_a_log_level) and find a
string in the output starting with `Using Kotlin Gradle plugin`, for example, `Using Kotlin Gradle plugin main variant`.

> Here are workarounds for some known issues with variant selection in Gradle:
> * [ResolutionStrategy in pluginManagement is not working for plugins with multivariants](https://github.com/gradle/gradle/issues/20545)
> * [Plugin variants are ignored when a plugin is added as the `buildSrc` common dependency](https://github.com/gradle/gradle/issues/20847)
>

> > {type="note"}

Reviewer: ```suggestion
{type="note"}
```

  initial step toward integrating Kotlin/Native build tools into all the other tools.

Please leave your feedback in [this YouTrack ticket](https://youtrack.jetbrains.com/issue/KT-32805).

### Statistics of generated files by each annotation processor in kapt

The `kotlin-kapt` Gradle plugin
already [reports performance statistics for each processor](https://github.com/JetBrains/kotlin/pull/4280). Starting
with Kotlin 1.7.0, it can also report statistics on the number of generated files for each annotation processor.

This is useful to track if there are unused annotation processors as a part of the build. You can use the generated
report to find modules that trigger unnecessary annotation processors and update the modules to prevent that.

Enable the statistics in two steps:

* Set the `showProcessorStats` flag to `true` in your `build.gradle.kts`:

  ```kotlin
  kapt {
      showProcessorStats = true
  }

* Set the `kapt.verbose` Gradle property to `true` in your `gradle.properties`:
  ```properties
  kapt.verbose=true
  ```

> You can also enable verbose output via the [command line option `verbose`](kapt.md#using-in-cli).
>

> > {type="note"}

Reviewer: ```suggestion
{type="note"}
```

#### Removal of the useExperimentalAnnotation method

In Kotlin 1.7.0, we completed the deprecation cycle for the `useExperimentalAnnotation` Gradle method. Use `optIn()`
instead to opt in to using an API in a module.

For example, if your Gradle module is multiplatform:

```kotlin
sourceSets {
    all {
        languageSettings.optIn("org.mylibrary.OptInAnnotation")
    }
}
```

Learn more about [opt-in requirements](opt-in-requirements.md) in Kotlin.

#### Removal of deprecated compiler options

We've completed the deprecation cycle for several compiler options:

* The `kotlinOptions.jdkHome` compiler option was deprecated in 1.5.30 and has been removed in the current release.
  Gradle builds now fail if they contain this option. We encourage you to
  use [Java toolchains](whatsnew1530.md#support-for-java-toolchains), which have been supported since Kotlin 1.5.30.
* The deprecated 'noStdlib' compiler option has also been removed. The Gradle plugin uses
  the `kotlin.stdlib.default.dependency=true` property to control whether the Kotlin standard library is present.

> The compiler arguments `-jdkHome` and `-no-stdlib` are still available.
>

> > {type="note"}

Reviewer: ```suggestion
{type="note"}
```

#### Removal of deprecated plugins

In Kotlin 1.4.0, the `kotlin2js` and `kotlin-dce-plugin` plugins were deprecated, and they have been removed in this
release. Instead of `kotlin2js`, use the new `org.jetbrains.kotlin.js` plugin. Dead code elimination (DCE) works when
the Kotlin/JS Gradle plugin is [properly configured](javascript-dce.md).

In Kotlin 1.6.0, we changed the deprecation level of the `KotlinGradleSubplugin` class to 'ERROR'. Developers used this
class for writing compiler plugins. In this
release, [this class has been removed](https://youtrack.jetbrains.com/issue/KT-48831/). Use
the `KotlinCompilerPluginSupportPlugin` class instead.

> The best practice is to use Kotlin plugins with versions 1.7.0 and higher throughout your project.
>

> > {type="tip"}

Reviewer: ```suggestion
{type="tip"}
```
